generic: 5.10 backport Treat IPv4 segment's lowest address as unicast

schoen opened the pull request:
Treat only the highest, not the lowest, IPv4 address within a local subnet as a broadcast address, as subnets do not need two different broadcast addresses and networking documentation consistently prefers the highest address as broadcast.

Backport of patch included in Linux 5.14.  Merged in upstream net-next tree in May 2021 at

https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=94c821c74bf5

This eventually frees up one address per subnet. It matches behavior suggested in our Internet-Draft, and also the default behavior of OpenBSD and FreeBSD.

Signed-off-by: Seth David Schoen <schoen@loyalty.org>

Ansuel commented:
@schoen pls refresh the patch and also fix the tag to something like generic: 5.10

Also I would suggest to generate the patch from a linux clone using git format-patch -1 HASH to have the correct patch name

schoen commented:
@Ansuel Thanks, is this more appropriate now?

schoen commented:
@Ansuel I see that some of these failures request that I run `make target/linux/refresh`, which I guess is what you were referring to with refreshing the patch. I just ran that, and it does a lot of work including applying patches in the tree, but the end result does not appear to be any different from git's point of view: after the `make target/linux/refresh`, `git status` still says:

`nothing to commit, working tree clean`

So I don't see how this step will affect the content of the pull request. Am I missing something?

Ansuel commented:
@schoen that command refresh generic AND current target patches. If any patch require refresh it will be changed in the related target dir.

Ansuel commented:
Anyway I'm running the checks to see what needs to be refreshed. Aside from that naming and patch looks to be correct!

Ansuel commented:
btw you should rebase this on top of master and drop the last commit (the merge one)

schoen commented:
I think I've got this appropriately rebased now, without the merge commit, thanks to some force-pushing on my branch.

schoen commented:
@Ansuel Could I ask you to please look at this again when you have a moment?

hauke commented:
Please refresh the patch like this:
````
make target/linux/{clean,refresh} V=99 -j5
````
It refreshes the patches from the currently selected target, select one which still uses kernel 5.10. 

I ran it on top of your branch and I got this:
````
$ git diff
diff --git a/target/linux/generic/backport-5.10/615-v5.14-ip-Treat-IPv4-segment-s-lowest-address-as-unicast.patch b/target/linux/generic/backport-5.10/615-v5.14-ip-Treat-IPv4-segment-s-lowest-address-as-unicast.patch
index 7267614e3573..8aabbe285994 100644
--- a/target/linux/generic/backport-5.10/615-v5.14-ip-Treat-IPv4-segment-s-lowest-address-as-unicast.patch
+++ b/target/linux/generic/backport-5.10/615-v5.14-ip-Treat-IPv4-segment-s-lowest-address-as-unicast.patch
@@ -16,11 +16,9 @@ Link: https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit
  net/ipv4/fib_frontend.c | 4 +---
  1 file changed, 1 insertion(+), 3 deletions(-)
 
-diff --git a/net/ipv4/fib_frontend.c b/net/ipv4/fib_frontend.c
-index 84bb707bd88d..bfb345c88271 100644
 --- a/net/ipv4/fib_frontend.c
 +++ b/net/ipv4/fib_frontend.c
-@@ -1122,10 +1122,8 @@ void fib_add_ifaddr(struct in_ifaddr *ifa)
+@@ -1129,10 +1129,8 @@ void fib_add_ifaddr(struct in_ifaddr *if
                                  prefix, ifa->ifa_prefixlen, prim,
                                  ifa->ifa_rt_priority);
  
@@ -31,7 +29,4 @@ index 84bb707bd88d..bfb345c88271 100644
 -                                prim, 0);
                        fib_magic(RTM_NEWROUTE, RTN_BROADCAST, prefix | ~mask,
                                  32, prim, 0);
-               }
--- 
-2.25.1
-
+                       arp_invalidate(dev, prefix | ~mask, false); 
````

schoen committed:
generic: 5.10: backport Treat IPv4 segment's lowest address as unicast

Backport patch from kernel 5.14.

Treat only the highest, not the lowest, IPv4 address within a local
subnet as a broadcast address, as subnets do not need two different
broadcast addresses and networking documentation consistently prefers
the highest address as broadcast.

This patch was merged in upstream net-next tree in May 2021 at

https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=94c821c74bf5

This eventually frees up one address per subnet. It matches behavior
suggested in our Internet-Draft, and also the default behavior of OpenBSD
and FreeBSD.

Signed-off-by: Seth David Schoen <schoen@loyalty.org>

schoen commented:
@hauke Thanks, I didn't understand what refreshing meant in this context; that did appear to make a difference in the diff (I guess a function argument was renamed in the upstream kernel afterward?).

This version that I pushed is now from after running that refresh command, so I hope it'll be correct.
